File: website/src/features/account-mapping/data/repositories/AccountMappingRepositoryImpl.ts
/**
 * AccountMappingRepositoryImpl
 * Implementation of IAccountMappingRepository interface
 *
 * Following Clean Architecture:
 * - Uses Model for data transformation (DTO ↔ Entity)
 * - Delegates data access to DataSource
 */

import {
  IAccountMappingRepository,
  AccountMappingResult,
  CreateMappingResult,
  DeleteMappingResult,
} from '../../domain/repositories/IAccountMappingRepository';
import { AccountMappingDataSource } from '../datasources/AccountMappingDataSource';
import { AccountMappingModel } from '../models/AccountMappingModel';

export class AccountMappingRepositoryImpl implements IAccountMappingRepository {
  private dataSource: AccountMappingDataSource;

  constructor() {
    this.dataSource = new AccountMappingDataSource();
  }

  async getAccountMappings(companyId: string): Promise<AccountMappingResult> {
    try {
      const data = await this.dataSource.getAccountMappings(companyId);

      // Use Model to parse RPC response and transform to domain entities
      const { allMappings } = AccountMappingModel.parseRPCResponse(data);

      return {
        success: true,
        data: allMappings,
      };
    } catch (error) {
      console.error('Repository error fetching account mappings:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to fetch account mappings',
      };
    }
  }

  async createAccountMapping(
    myCompanyId: string,
    counterpartyCompanyId: string,
    myAccountId: string,
    linkedAccountId: string,
    direction: string,
    createdBy: string
  ): Promise<CreateMappingResult> {
    try {
      await this.dataSource.createAccountMapping(
        myCompanyId,
        counterpartyCompanyId,
        myAccountId,
        linkedAccountId,
        direction,
        createdBy
      );

      return {
        success: true,
        data: undefined,
      };
    } catch (error) {
      console.error('Repository error creating account mapping:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to create account mapping',
      };
    }
  }

  async getCompanyAccounts(companyId: string) {
    try {
      const data = await this.dataSource.getCompanyAccounts(companyId);

      // Use Model to extract accounts from RPC response
      const accounts = AccountMappingModel.extractAccounts(data);

      return {
        success: true,
        data: accounts,
      };
    } catch (error) {
      console.error('Repository error fetching company accounts:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to fetch company accounts',
      };
    }
  }

  async deleteAccountMapping(mappingId: string): Promise<DeleteMappingResult> {
    try {
      await this.dataSource.deleteAccountMapping(mappingId);

      return {
        success: true,
      };
    } catch (error) {
      console.error('Repository error deleting account mapping:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to delete account mapping',
      };
    }
  }
}

File: myFinance_improved_V2/lib/features/counter_party/data/repositories/account_mapping_repository_impl.dart
import '../../domain/entities/account_mapping.dart';
import '../../domain/repositories/account_mapping_repository.dart';
import '../datasources/account_mapping_datasource.dart';
import '../models/account_mapping_dto.dart';

/// Account Mapping Repository Implementation
class AccountMappingRepositoryImpl implements AccountMappingRepository {
  final AccountMappingDataSource _dataSource;

  AccountMappingRepositoryImpl(this._dataSource);

  @override
  Future<List<AccountMapping>> getAccountMappings({
    required String counterpartyId,
  }) async {
    try {
      final data = await _dataSource.getAccountMappings(
        counterpartyId: counterpartyId,
      );

      return data
          .map((json) => AccountMappingDto.fromJson(json).toEntity())
          .toList();
    } catch (e) {
      throw Exception('Failed to load account mappings: $e');
    }
  }

  @override
  Future<AccountMapping> createAccountMapping({
    required String myCompanyId,
    required String myAccountId,
    required String counterpartyId,
    required String linkedAccountId,
    String direction = 'bidirectional',
    String? createdBy,
  }) async {
    try {
      final data = await _dataSource.createAccountMapping(
        myCompanyId: myCompanyId,
        myAccountId: myAccountId,
        counterpartyId: counterpartyId,
        linkedAccountId: linkedAccountId,
        direction: direction,
        createdBy: createdBy,
      );

      final mappingType = data['type'] as String?;

      // Fetch the complete mapping data (always ordered by created_at DESC from RPC)
      final mappings = await getAccountMappings(counterpartyId: counterpartyId);

      if (mappings.isEmpty) {
        throw Exception('Created mapping not found after insert');
      }

      // Handle different response types
      if (mappingType == 'single_direction') {
        // Single direction: Internal transaction, one mapping created
        final mappingId = data['mapping_id'] as String;
        final newMapping = mappings.firstWhere(
          (m) => m.mappingId == mappingId,
          orElse: () => throw Exception('Created mapping not found with ID: $mappingId'),
        );
        return newMapping;
      } else if (mappingType == 'bidirectional') {
        // Bidirectional: Cross-company transaction, two mappings created
        // Return the most recent mapping (first in list since ordered by created_at DESC)
        return mappings.first;
      } else {
        // Fallback: try both methods for backward compatibility
        if (data.containsKey('mapping_id')) {
          final mappingId = data['mapping_id'] as String;
          final newMapping = mappings.firstWhere(
            (m) => m.mappingId == mappingId,
            orElse: () => mappings.first,
          );
          return newMapping;
        } else {
          return mappings.first;
        }
      }
    } catch (e) {
      // Provide clear error messages to the user
      final errorMsg = e.toString();
      if (errorMsg.contains('Duplicate mapping')) {
        throw Exception('This account mapping already exists for this counterparty');
      } else if (errorMsg.contains('Counterparty does not belong')) {
        throw Exception('Invalid counterparty: This counterparty belongs to a different company');
      } else if (errorMsg.contains('unique_violation') || errorMsg.contains('Database constraint')) {
        throw Exception('Cannot create mapping: This would create a duplicate entry');
      }
      throw Exception('Failed to create account mapping: $e');
    }
  }

  @override
  Future<bool> deleteAccountMapping({
    required String mappingId,
  }) async {
    return await _dataSource.deleteAccountMapping(
      mappingId: mappingId,
    );
  }

  @override
  Future<List<Map<String, dynamic>>> getAvailableAccounts({
    required String companyId,
  }) async {
    return await _dataSource.getAvailableAccounts(
      companyId: companyId,
    );
  }

  @override
  Future<Map<String, dynamic>?> getLinkedCompanyInfo({
    required String counterpartyId,
  }) async {
    return await _dataSource.getLinkedCompanyInfo(
      counterpartyId: counterpartyId,
    );
  }
}
